nutrition_labels/grant_data_processing.py

```diff
@@ -0,0 +1,119 @@
+        epmc_list.append({'pmid':row['pmid'],
+                          'code':row['code'],
+                          'grant_number':num})
+
+epmc_df = pd.DataFrame(epmc_list)
+epmc_df = epmc_df[epmc_df['code'].isin([1,2,3])] # Only selecting useful codes
+
+# getting WT grant number
+def remove_useless_string(string):
+    '''
+    cleans the grant descriptions of artifacts such as <br />
+    :param string: description string
+    :return: clean string
+    '''
+
+    soup = BeautifulSoup(string)
+    string_out = soup.get_text()
+    string_out = string_out.strip('\n')
+    string_out = string_out.strip('\xa0')
+    string_out = re.sub('  ','',string_out)
+    return(string_out)
+
+def only_text(string):
+    '''
+    removes non-alphanumeric characters and spaces to increase matching
+    :param string: description string
+    :return: clean string
+    '''
+    string = re.sub(' |\W','',string)
+    return(string)
+
+
+grant_ref = grant_data[['Grant Programme:Title','Internal ID','Description','Award Date']]
+grant_ref['grant_number'] = grant_data['Internal ID'].apply(lambda x: re.sub('/.*','',x))
+grant_ref = pd.merge(grant_ref,epmc_df, how = 'inner', on = 'grant_number')
+grant_ref['Description'] = grant_ref['Description'].apply(remove_useless_string)
+grant_ref['Description'] = grant_ref['Description'].apply(only_text)
+grant_ref = grant_ref.drop_duplicates(subset=['Description','grant_number','pmid'])
+
+# get remaining entrys with the same pmid grant number but different description
+grant_duplicates = grant_ref[grant_ref.duplicated(subset=['pmid','grant_number'], keep = False)]
+remove_rows = grant_duplicates.index.tolist()
+
+# sort remove list from duplicates (the later grants)
+grant_duplicates = grant_duplicates.sort_values('Award Date').drop_duplicates(subset = ['grant_number','pmid'])
+remove_rows = [i for i in remove_rows if i not in grant_duplicates.index.tolist()]
+grant_ref = grant_ref.drop(remove_rows)
+
+# finding abstract without a matching code
+no_ref = epmc_df[~epmc_df['grant_number'].isin(grant_ref['grant_number'].to_list())]
+no_ref = no_ref[~no_ref['pmid'].isin(grant_ref['pmid'].to_list())]
+missing_abstracts = epmc_tags[epmc_tags['pmid'].isin(no_ref['pmid'].to_list())] #I cant find where this grant is at all
+
+# get list
+code_list = grant_ref[['Internal ID','code','pmid']]
+
+# get rf data
+rf_useful = rf_tags[rf_tags['code '].isin([1,2,3])]
+rf_useful = rf_useful[['code ','Grant Reference','Name']]
+rf_useful = rf_useful.rename({'code ':'code','Grant Reference':'Internal ID','Name':'pmid'}, axis = 1)
+
+# merge abstract and rfs
+code_list = pd.concat([code_list,rf_useful])
+code_list.loc[:,'Internal ID'] = code_list['Internal ID'].str.strip(' ')
+code_list_final = pd.merge(grant_data[['Internal ID','Title','Description']],code_list, how = 'inner', on = 'Internal ID')
+rf_not_in = code_list[~code_list['Internal ID'].isin(code_list_final['Internal ID'])] # six codes arent in the wellcome data
```

Comment:
again here I'd do a print of the results, and maybe just the grant IDs are the interesting bit
```
print("Grants flagged as relevant in RF data not found in grants data: ")
print(set(code_list['Internal ID']).difference(set(code_list_final['Internal ID'])))
```